src/lem/constrainer.rs

```diff
@@ -81,145 +117,320 @@ impl LEM {
         Ok(())
     }
 
-    /// Create R1CS constraints for LEM given an evaluation witness.
+    fn on_concrete_path(concrete_path: &Boolean) -> Result<bool> {
+        concrete_path
+            .get_value()
+            .ok_or_else(|| anyhow!("Couldn't check whether we're on a concrete path"))
+    }
+
+    fn z_ptr_from_valuation<F: LurkField>(
+        concrete_path: &Boolean,
+        valuation: &Valuation<F>,
+        name: &String,
+        store: &mut Store<F>,
+    ) -> Result<ZPtr<F>> {
+        if Self::on_concrete_path(concrete_path)? {
+            let Some(ptr) = valuation.ptrs.get(name) else {
+                bail!("Couldn't retrieve {} from valuation", name);
+            };
+            store.hydrate_ptr(ptr)
+        } else {
+            Ok(ZPtr::dummy())
+        }
+    }
+
+    /// Accumulates slot data that will be used later to generate the constraints.
+    fn acc_slots_data_data<F: LurkField>(
+        concrete_path: Boolean,
+        slots_data: &mut SlotsData<F>,
+        img: AllocatedNum<F>,
+        img_name: String,
+        preimg_vec: Vec<&AllocatedPtr<F>>,
+    ) {
+        let arity = match preimg_vec.len() {
+            2 => HashArity::A2,
+            3 => HashArity::A3,
+            4 => HashArity::A4,
+            _ => unreachable!(),
+        };
+        let mut input_vec = Vec::new();
+        for elem in preimg_vec {
+            input_vec.push(elem.tag().clone());
+            input_vec.push(elem.hash().clone());
+        }
+        slots_data.push((arity, concrete_path, input_vec, img, img_name));
+    }
+
+    /// Use the implies logic to contrain tag and hash values for accumulated
+    /// slot data
+    fn create_slot_constraints<F: LurkField, CS: ConstraintSystem<F>>(
+        cs: &mut CS,
+        slots_data: &SlotsData<F>,
+        store: &mut Store<F>,
+        alloc_manager: &mut AllocationManager<F>,
+        num_hash_slots: &NumSlots,
+    ) -> Result<()> {
+        // Vectors fulls of dummies, so that it will not be required to fill with dummies later
+        let alloc_dummy_ptr = alloc_manager.get_or_alloc_ptr(cs, &ZPtr::dummy())?;
+        let mut hashes = vec![Some(alloc_dummy_ptr.hash().clone()); num_hash_slots.total()];
+
+        // In order to get a uniform circuit each type of hash has its own constant-size subvector
+        // Hash2 uses subvector from 0 to num_hash_slots.hash2
+        let mut hash2_index = 0;
+        // Hash3 uses subvector from num_hash_slots.hash2 to num_hash+slots.hash3
+        let mut hash3_index = num_hash_slots.hash2;
+        // Hash4 uses subvector from num_hash_slots.hash3 to num_hash+slots.hash4
+        let mut hash4_index = num_hash_slots.hash3;
+        for (arity, concrete_path, preimg_vec, img, img_name) in slots_data.iter() {
+            let is_concrete_path = Self::on_concrete_path(concrete_path)?;
+            if is_concrete_path {
+                macro_rules! constrain_slot {
```

Comment:
If we add hash constraints independently of the value of this boolean, then we would not have the optimization. On the other hand I see what you are pointing out here. We have constrained dummy implications, which constrains the result of the hash to be equal to the allocated output pointer.  But we are not constraining the dummy hashes many times, since I was considering this a first example of a deduplication of pointers. But you are right. the circuit size would change depending on the path followed. To maintain uniformity we need many dummy hashes indeed.

Reply:
I've made a slight change in the code to improve readability. But meaning is kept.

Reply:
Yes, @emmorais. The whole point of dummy hashes is that they stand in for hashes that would happen otherwise. The circuit mechanism for checking whether a hash is real or dummy needs to be identical in all cases. It's just that depending on the path taken, the *result* of the computation (whether dummy or not) will be different. Nothing about the shape of the circuit and its computation can differ though.

Unless you are telling me this code has the property I just described, then I think the meaning needs to be changed. If that is the case, I'm a little surprised, since I think this is one of the main points we've been discussing all along. This is , in effect, the entire point of the implication+dummy strategy we decided to lean into.